shell/platform/android/test/io/flutter/plugin/editing/InputConnectionAdaptorTest.java

```diff
@@ -341,6 +341,33 @@ public void testSendKeyEvent_rightKeyMovesCaretRight() {
     assertEquals(selStart + 1, Selection.getSelectionEnd(editable));
   }
 
+  @Test
+  public void testSendKeyEvent_rightKeyMovesCaretRightComplexRegion() {
+    int selStart = 0;
+    // Seven region indicator characters. The first six should be considered as
+    // three region indicators, and the final seventh character should be
+    // considered to be on its own because it has no partner.
+    String SAMPLE_REGION_TEXT = "🇷🇷🇷🇷🇷🇷🇷";
+    Editable editable = sampleEditable(selStart, selStart, SAMPLE_REGION_TEXT);
+    InputConnectionAdaptor adaptor = sampleInputConnectionAdaptor(editable);
+
+    KeyEvent downKeyDown = new KeyEvent(KeyEvent.ACTION_DOWN, KeyEvent.KEYCODE_DPAD_RIGHT);
+    boolean didConsume;
+
+    didConsume = adaptor.sendKeyEvent(downKeyDown);
+    assertTrue(didConsume);
+    assertEquals(Selection.getSelectionStart(editable), 4);
+    didConsume = adaptor.sendKeyEvent(downKeyDown);
+    assertTrue(didConsume);
+    assertEquals(Selection.getSelectionStart(editable), 8);
```

Comment:
I think it's good to add another situation for this test such that if we are in offset 6, then the right key should bring us to offset 8.

Reply:
Good call.  Added in 1c5df01.